# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: modules/libpref/Preferences.cpp
# Commit: d0d2fe421be4
# Full Hash: d0d2fe421be4dcdfa4d503be79046dbe69cdf520
# Author: Jan Horak <jhorak@redhat.com>
# Date: 2023-02-21 21:26:48
# Regressor Bug: 1170092
# File Overlap Count: 1
# Description:
#   Bug 1170092 Load preferences from /etc/firefox/defaults/pref on Linux; r=mkaply
#   
#   Preferences from /etc/firefox/defaults/pref overwrites preferences set in omni.jar
#   and firefox/defaults/pref. The code was used for flatpak and snap only, but it could
#   be used also for regular Linux builds.
# ==============================================================================

diff -r be618bd7d68d -r d0d2fe421be4 modules/libpref/Preferences.cpp
--- a/modules/libpref/Preferences.cpp	Tue Feb 21 14:10:39 2023 +0000
+++ b/modules/libpref/Preferences.cpp	Tue Feb 21 14:12:42 2023 +0000
@@ -4921,23 +4921,6 @@
     NS_WARNING("Error parsing application default preferences.");
   }
 
-#if defined(MOZ_WIDGET_GTK)
-  // Under Flatpak/Snap package, load /etc/firefox/defaults/pref/*.js.
-  if (mozilla::widget::IsRunningUnderFlatpakOrSnap()) {
-    nsCOMPtr<nsIFile> defaultSnapPrefDir;
-    rv = NS_GetSpecialDirectory(NS_OS_SYSTEM_CONFIG_DIR,
-                                getter_AddRefs(defaultSnapPrefDir));
-    NS_ENSURE_SUCCESS(rv, rv);
-    defaultSnapPrefDir->AppendNative("defaults"_ns);
-    defaultSnapPrefDir->AppendNative("pref"_ns);
-
-    rv = pref_LoadPrefsInDir(defaultSnapPrefDir, nullptr, 0);
-    if (NS_FAILED(rv)) {
-      NS_WARNING("Error parsing application default preferences under Snap.");
-    }
-  }
-#endif
-
   // Load jar:$app/omni.jar!/defaults/preferences/*.js
   // or jar:$gre/omni.jar!/defaults/preferences/*.js.
   RefPtr<nsZipArchive> appJarReader = Omnijar::GetReader(Omnijar::APP);
@@ -5010,6 +4993,24 @@
     }
   }
 
+#if defined(MOZ_WIDGET_GTK)
+  // To ensure the system-wide preferences are not overwritten by
+  // firefox/browser/defauts/preferences/*.js we need to load
+  // the /etc/firefox/defaults/pref/*.js settings as last.
+  // Under Flatpak, the NS_OS_SYSTEM_CONFIG_DIR points to /app/etc/firefox
+  nsCOMPtr<nsIFile> defaultSystemPrefDir;
+  rv = NS_GetSpecialDirectory(NS_OS_SYSTEM_CONFIG_DIR,
+                              getter_AddRefs(defaultSystemPrefDir));
+  NS_ENSURE_SUCCESS(rv, rv);
+  defaultSystemPrefDir->AppendNative("defaults"_ns);
+  defaultSystemPrefDir->AppendNative("pref"_ns);
+
+  rv = pref_LoadPrefsInDir(defaultSystemPrefDir, nullptr, 0);
+  if (NS_FAILED(rv)) {
+    NS_WARNING("Error parsing application default preferences.");
+  }
+#endif
+
   if (XRE_IsParentProcess()) {
     SetupTelemetryPref();
   }
